x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +47,146 @@ func newRetryLogger() *retryLogger {
+	log := ctx.Logger.With("url", in.config.URL)
 
-	var dateStr string
-	if in.nextCursorValue == "" {
-		t := timeNow().UTC().Add(-in.config.DateCursor.InitialInterval)
-		dateStr = t.Format(in.config.DateCursor.GetDateFormat())
-	} else {
-		dateStr = in.nextCursorValue
-	}
+	stdCtx := ctxtool.FromCanceller(ctx.Cancelation)
 
-	url, err := url.Parse(in.config.URL)
+	httpClient, err := in.newHTTPClient(stdCtx)
 	if err != nil {
-		return in.config.URL
-	}
-
-	q := url.Query()
-
-	var value string
-	if in.config.DateCursor.ValueTemplate == nil {
-		value = dateStr
-	} else {
-		buf := new(bytes.Buffer)
-		if err := in.config.DateCursor.ValueTemplate.Execute(buf, dateStr); err != nil {
-			return in.config.URL
-		}
-		value = buf.String()
+		return err
 	}
 
-	q.Set(in.config.DateCursor.URLField, value)
-
-	url.RawQuery = q.Encode()
-
-	return url.String()
-}
-
-func (in *HttpjsonInput) advanceCursor(m common.MapStr) {
-	if in.config.DateCursor.Field == "" {
-		in.nextCursorValue = time.Now().UTC().Format(in.config.DateCursor.GetDateFormat())
-		return
-	}
+	dateCursor := newDateCursorFromConfig(in.config, log)
 
-	v, err := m.GetValue(in.config.DateCursor.Field)
-	if err != nil {
-		in.log.Warnf("date_cursor field: %q", err)
-		return
-	}
-	switch t := v.(type) {
-	case string:
-		_, err := time.Parse(in.config.DateCursor.GetDateFormat(), t)
-		if err != nil {
-			in.log.Warn("date_cursor field does not have the expected layout")
-			return
-		}
-		in.nextCursorValue = t
-	default:
-		in.log.Warn("date_cursor field must be a string, cursor will not advance")
-		return
-	}
-}
+	rateLimiter := newRateLimiterFromConfig(in.config, log)
 
-func (in *HttpjsonInput) run() error {
-	ctx, cancel := context.WithCancel(in.workerCtx)
-	defer cancel()
+	pagination := newPaginationFromConfig(in.config)
 
-	client, err := in.newHTTPClient(ctx)
-	if err != nil {
-		return err
-	}
+	requester := newRequester(
+		in.config,
+		rateLimiter,
+		dateCursor,
+		pagination,
+		httpClient,
+		log,
+	)
 
-	ri := &RequestInfo{
-		ContentMap: common.MapStr{},
-		Headers:    in.HTTPHeaders,
-	}
-	if in.config.HTTPMethod == "POST" && in.config.HTTPRequestBody != nil {
-		ri.ContentMap.Update(common.MapStr(in.config.HTTPRequestBody))
-	}
-	err = in.processHTTPRequest(ctx, client, ri)
-	if err == nil && in.Interval > 0 {
-		ticker := time.NewTicker(in.Interval)
+	err = requester.processHTTPRequest(stdCtx, publisher)
+	if err == nil && in.config.Interval > 0 {
+		ticker := time.NewTicker(in.config.Interval)
 		defer ticker.Stop()
 		for {
 			select {
-			case <-ctx.Done():
-				in.log.Info("Context done.")
+			case <-stdCtx.Done():
+				log.Info("Context done.")
 				return nil
 			case <-ticker.C:
-				in.log.Info("Process another repeated request.")
-				err = in.processHTTPRequest(ctx, client, ri)
+				log.Info("Process another repeated request.")
+				err = requester.processHTTPRequest(stdCtx, publisher)
 				if err != nil {
 					return err
 				}
 			}
 		}
 	}
```

Comment:
This loop is quite a common pattern. Better use `timed.Periodic(stdCtx, in.Config.Interval, func() { ... })`. Using this pattern to check for cancellation in a loop one should add an extra select on the done channel only in order to confirm done has not been closed before doing the actual select statement.

Periodic is defined here: https://github.com/elastic/go-concert/blob/master/timed/timed.go#L57